packages/camera/camera_android_camerax/lib/src/live_camera_state.dart

```diff
@@ -0,0 +1,117 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/services.dart';
+
+import 'android_camera_camerax_flutter_api_impls.dart';
+import 'camerax_library.g.dart';
+import 'instance_manager.dart';
+import 'java_object.dart';
+
+/// The live state of the camera.
+///
+/// This is the LiveData of type CameraState that Android tracks in relation
+/// to a CameraInfo instance. See
+/// https://developer.android.com/reference/androidx/camera/core/CameraInfo#getCameraState().
+class LiveCameraState extends JavaObject {
+  /// Creates a detached [LiveCameraState].
+  LiveCameraState.detached({
+    BinaryMessenger? binaryMessenger,
+    InstanceManager? instanceManager,
+  }) : super.detached(
+            binaryMessenger: binaryMessenger,
+            instanceManager: instanceManager) {
+    _api = LiveCameraStateHostApiImpl(
+        binaryMessenger: binaryMessenger, instanceManager: instanceManager);
+    AndroidCameraXCameraFlutterApis.instance.ensureSetUp();
+  }
+
+  /// Stream that emits an event when the related [Camera] instance starts
+  /// to close.
+  static final StreamController<bool> cameraClosingStreamController =
```

Comment:
Are you using this for callback methods of the `Observer`? What happens if there are two `LiveCameraState` that are instantiated? How would you be able to tell which instance is receiving a callback?

Reply:
That is true. We really should avoid that, but we should be prepared for that, just in case. My first thought would be comparing the identifiers of the `Observer`'s caller and the `LiveCameraState` instance, but this would mean we expose the `InstanceManager` to actual plugin code, which I don't think is desired...or maybe we can try ensuring there is only one instance or making that clear in the plugin code?

Reply:
From my understanding, you could have a camera state for the front and back camera, right? For `webview_flutter` we handle callback methods by passing a Function to the constructor: https://github.com/flutter/packages/blob/main/packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit/web_kit.dart#L821.

This is the main reason the `InstanceManager` requires an `onCopy` method. Copied instances have the same callback functions as the original. 

Then in the `FlutterApi` we check if the function is nonnull and call it: https://github.com/flutter/packages/blob/main/packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit/web_kit_api_impls.dart#L763

Reply:
However, it looks like you would have to wrap the `Observer` class to do this and pass it to `addObserver`. Let me know if you're able to use the generator. I've made improvements and it should be really good at creating the code for this.

Reply:
Two points:

1. For every `Camera` you can get a `CameraInfo` instance, from which you can get a `LiveData<CameraState>` instance. So, if we want to add callbacks to a constructor to understand which Java `LiveData<CameraState>`/Dart `LiveCameraState` is calling the `onCameraClosing()` callback method, I don't see why we can't just modify the Dart `LiveCameraState` constructor directly (or `CameraInfo`/`Camera` if not) versus needing to wrap `Observer`.

2. I do think adding a callback to one of the classes could be helpful for clarity because I do understand the concern of not knowing which Java `LiveData<CameraState>`/Dart `LiveCameraState` is calling `onCameraClosing()`, but I'm not sure how this would necessarily mitigate the effects of that. If for some reason, the plugin tracks multiple `Camera`s and multiple `CameraInfo`s/`LiveCameraState`s are created, then they would both end up adding events to the stream returned by `onCameraClosing()` anyways. However, the plugin should not do that. It really only tracks one `Camera` at a time. Maybe we need a way to enforce that on top of adding the callback.

Reply:
We decided to move forward with wrapping all of the classes. It seems like there may be an impact on performance, as I noticed that by wrapping all of the classes, I was able to receive camera states but not the closing state, whereas with the old way, I was able to see the closing state, as well. We suspect this is due to the warning noted here: https://developer.android.com/reference/androidx/camera/core/CameraInfo#getCameraState%28%29.

Despite this, I decided that wrapping the classes was the correct way to move forward to commit to the approach of wrapping native classes in Dart. Next steps could be filing a bug for this to track the issue, writing an integration test to test this in a device lab, and following up with CameraX folks on the issue.